### PR TITLE
Modify alter tablet implementation and change ColumnPruning to ZoneMap

### PR DESCRIPTION
**Pull request by chaoyli**
1. Unify alter table state resided in memory and disk.
2. Merge the two process_alter_tablet and _do_alter_tablet function.
3. Rename _alter_tablet to _convert_historical_rowsets.
4. Rename ColumnPruning to ZoneMap.

### CODE REVIEW

File: be/src/olap/cumulative_compaction.cpp [hunk 1]
```diff
@@ -452,9 +453,6 @@ OLAPStatus CumulativeCompaction::_do_cumulative_compaction() {
     // 6. delete delta files which have been merged into new cumulative file
     _delete_unused_delta_files(&unused_rowsets);
```
Comment (kangpinghuang) [inline be/src/olap/cumulative_compaction.cpp]: rename this function name to _delete_unused_rowset?

File: be/src/olap/base_compaction.cpp [hunk 2]
```diff
@@ -174,7 +172,10 @@ OLAPStatus BaseCompaction::run() {
     _release_base_compaction_lock();
 
     LOG(INFO) << "succeed to do base compaction. tablet=" << _tablet->full_name()
-              << ", base_version=" << _new_base_version.first << "-" << _new_base_version.second;
+              << ", base_version=" << _new_base_version.first << "-" << _new_base_version.second
+              << ". elapsed time of doing base compaction"
+              << ", time=" << stage_watch.get_elapse_time_us() / (100000.0) << "s";
```
Comment (kangpinghuang) [inline be/src/olap/base_compaction.cpp]: I think there is need to  add a get_elapse_time_s() function?

File: be/src/olap/olap_cond.cpp [hunk 3]
```diff
@@ -601,25 +601,24 @@ bool Conditions::delete_conditions_eval(const RowCursor& row) const {
 }
 
 bool Conditions::delta_pruning_filter(
```
Comment (kangpinghuang) [inline be/src/olap/olap_cond.cpp]: rowset_pruning_filter?

File: be/src/olap/olap_cond.cpp [hunk 4]
```diff
@@ -601,25 +601,24 @@ bool Conditions::delete_conditions_eval(const RowCursor& row) const {
...
-        const std::vector<std::pair<WrapperField*, WrapperField*>>& column_statistics) const {
+        const std::vector<std::pair<WrapperField*, WrapperField*>>& zone_maps) const {
```
Comment (kangpinghuang) [inline be/src/olap/olap_cond.cpp]: use KeyRange to confirm to the statement of this function

File: be/src/olap/olap_cond.cpp [hunk 5]
```diff
@@ -601,25 +601,24 @@ bool Conditions::delete_conditions_eval(const RowCursor& row) const {
...
+        if (cond_it.second->is_key() && !cond_it.second->eval(zone_maps[cond_it.first])) {
             return true;
         }
     }
     return false;
 }
 
 int Conditions::delete_pruning_filter(
-        const std::vector<std::pair<WrapperField*, WrapperField*>>& col_stat) const {
+        const std::vector<std::pair<WrapperField*, WrapperField*>>& zone_maps) const {
```
Comment (kangpinghuang) [inline be/src/olap/olap_cond.cpp]: Use KeyRange

File: be/src/olap/olap_cond.cpp [hunk 6]
```diff
@@ -656,7 +654,7 @@ int Conditions::delete_pruning_filter(
         }
     }
 
-    if (true == del_not_satisfied || 0 == _columns.size()) {
+    if (del_not_satisfied || _columns.empty()) {
```
Comment (kangpinghuang) [inline be/src/olap/olap_cond.cpp]: put _columns.empty() in the first line of this function

File: be/src/olap/rowset/column_data_writer.cpp [hunk 7]
```diff
@@ -197,7 +196,7 @@ OLAPStatus ColumnDataWriter::finalize() {
         return res;
     }
 
-    res = _segment_group->add_column_statistics(_column_statistics);
+    res = _segment_group->add_zone_maps(_zone_maps);
     if (res != OLAP_SUCCESS) {
         OLAP_LOG_WARNING("Fail to set delta pruning![res=%d]", res);
```
Comment (kangpinghuang) [inline be/src/olap/rowset/column_data_writer.cpp]: LOG(WARNING)

File: be/src/olap/schema_change.h [hunk 8]
```diff
@@ -268,7 +268,7 @@ class SchemaChangeHandler {
     OLAPStatus process_alter_tablet(AlterTabletType alter_tablet_type,
                                    const TAlterTabletReq& request);
 
-    OLAPStatus schema_version_convert(TabletSharedPtr ref_tablet,
+    OLAPStatus schema_version_convert(TabletSharedPtr base_tablet,
```
Comment (kangpinghuang) [inline be/src/olap/schema_change.h]: call this function in delta_writer

File: be/src/olap/schema_change.cpp [hunk 9]
```diff
@@ -1172,100 +1172,65 @@ OLAPStatus SchemaChangeHandler::_check_and_clear_schema_change_info(
...
 
     // 4. return failed if new tablet already exist in StorageEngine.
     TabletSharedPtr new_tablet = TabletManager::instance()->get_tablet(
             request.new_tablet_req.tablet_id, request.new_tablet_req.tablet_schema.schema_hash);
-    if (new_tablet.get() != NULL) {
-        res = OLAP_SUCCESS;
-    } else {
-        res = _do_alter_tablet(type, ref_tablet, request);
+    if (new_tablet != nullptr) {
+        TabletManager::instance()->release_schema_change_lock(request.base_tablet_id);
```
Comment (kangpinghuang) [inline be/src/olap/schema_change.cpp]: where to take the new tablet schema change lock before?
this place should not unlock

File: be/src/olap/schema_change.cpp [hunk 10]
```diff
@@ -1279,56 +1244,48 @@ OLAPStatus SchemaChangeHandler::_do_alter_tablet(
...
                     << ", transaction_id=" << transaction_id
                     << ", transaction_size=" << transaction_ids.size();
         }
     }
 
     // 2. Get version_to_be_changed and store into tablet header
-    ref_tablet->obtain_push_lock();
-    ref_tablet->obtain_header_wrlock();
+    base_tablet->obtain_push_lock();
+    base_tablet->obtain_header_wrlock();
```
Comment (kangpinghuang) [inline be/src/olap/schema_change.cpp]: this name should be rename to obtain_meta_wrlock?
header should not be used

File: be/src/olap/schema_change.cpp [hunk 11]
```diff
@@ -46,12 +46,12 @@ using std::vector;
 namespace doris {
```
Comment (kangpinghuang) [inline be/src/olap/schema_change.cpp]: This file should be rename to alter_tablet_handler.cpp?

File: be/src/olap/schema_change.cpp [hunk 12]
```diff
@@ -1488,8 +1419,8 @@ OLAPStatus SchemaChangeHandler::_create_new_tablet(
                     << ", tablet=" << request.tablet_id
                     << ":" << request.tablet_schema.schema_hash;
             }
-        } else if (NULL != new_tablet) {
-            new_tablet->delete_all_files();
+        } else if (NULL != tablet_to_create) {
```
Comment (kangpinghuang) [inline be/src/olap/schema_change.cpp]: tablet_to_create != nullptr

File: be/src/olap/schema_change.cpp [hunk 13]
```diff
@@ -1632,166 +1563,156 @@ OLAPStatus SchemaChangeHandler::schema_version_convert(
...
-    OLAPStatus res = OLAP_SUCCESS;
-
     // 1. 在新表和旧表中添加schema change标志
-    ref_tablet->delete_alter_task();
-    ref_tablet->add_alter_task(new_tablet->tablet_id(),
-                               new_tablet->schema_hash(),
-                               versions_to_be_changed,
-                               alter_tablet_type);
+    base_tablet->delete_alter_task();
+    base_tablet->add_alter_task(new_tablet->tablet_id(),
```
Comment (kangpinghuang) [inline be/src/olap/schema_change.cpp]: there is bug in add_alter_task function, pls check it.

File: be/src/olap/schema_change.cpp [hunk 14]
```diff
@@ -1632,166 +1563,156 @@ OLAPStatus SchemaChangeHandler::schema_version_convert(
...
+    base_tablet->set_alter_state(state);
+    OLAPStatus res = base_tablet->save_meta();
+    if (res != OLAP_SUCCESS) {
+        LOG(FATAL) << "fail to save ref tablet header. res=" << res
+                   << ", tablet=" << base_tablet->full_name();
+        return res;
+    }
+
+    new_tablet->set_alter_state(state);
+    res = new_tablet->save_meta();
```
Comment (kangpinghuang) [inline be/src/olap/schema_change.cpp]: redundant codes!!!

**Review by kangpinghuang — CHANGES_REQUESTED** [review]

File: be/src/olap/schema_change.cpp (hunk 14 above)
Comment (chaoyli) [reply be/src/olap/schema_change.cpp]: OK

File: be/src/olap/schema_change.cpp (hunk 13 above)
Comment (chaoyli) [reply be/src/olap/schema_change.cpp]: OK

File: be/src/olap/schema_change.cpp (hunk 12 above)
Comment (chaoyli) [reply be/src/olap/schema_change.cpp]: OK

File: be/src/olap/schema_change.cpp (hunk 9 above)
Comment (chaoyli) [reply be/src/olap/schema_change.cpp]: Only call try_schema_change_lock() of base_tablet.

File: be/src/olap/rowset/column_data_writer.cpp (hunk 7 above)
Comment (chaoyli) [reply be/src/olap/rowset/column_data_writer.cpp]: OK

File: be/src/olap/olap_cond.cpp (hunk 6 above)
Comment (chaoyli) [reply be/src/olap/olap_cond.cpp]: OK

File: be/src/olap/base_compaction.cpp (hunk 2 above)
Comment (chaoyli) [reply be/src/olap/base_compaction.cpp]: OK

File: be/src/olap/cumulative_compaction.cpp (hunk 1 above)
Comment (chaoyli) [reply be/src/olap/cumulative_compaction.cpp]: OK

File: be/src/olap/olap_cond.cpp (hunk 4 above)
Comment (chaoyli) [reply be/src/olap/olap_cond.cpp]: OK

File: be/src/olap/olap_cond.cpp (hunk 5 above)
Comment (chaoyli) [reply be/src/olap/olap_cond.cpp]: OK

File: be/src/olap/schema_change.cpp (hunk 11 above)
Comment (chaoyli) [reply be/src/olap/schema_change.cpp]: OK, this modification may refer to many codes, I will push it on a lonely commit.

File: be/src/olap/schema_change.cpp (hunk 10 above)
Comment (chaoyli) [reply be/src/olap/schema_change.cpp]: OK, this modification may refer to many codes, I will push it on a lonely commit.

File: be/src/olap/schema_change.h (hunk 8 above)
Comment (chaoyli) [reply be/src/olap/schema_change.h]: I will change this name in a following commit. So call it in the next commit.

File: be/src/olap/olap_cond.cpp (hunk 3 above)
Comment (chaoyli) [reply be/src/olap/olap_cond.cpp]: OK

**Review by kangpinghuang — APPROVED** [review]

**Pull request merged**